### ISSUE TITLE
"named existentials" in GADT patterns should support naming all type variables used in the constructor declaration

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#11891]
(This issue is extracted from a suggestion by 'kantian' in [this Discuss thread](https://discuss.ocaml.org/t/explicitly-name-types-in-gadt-patterns/11168/4?u=gasche))

Consider the following example (OCaml 5.0):

```ocaml
type _ th =
  | Thunk : 'a * ('a -> 'b) -> 'b th

(* this works *)
let f (type a) : a th -> a = function
  | Thunk (type b) (x, f : b * (b -> _)) -> f x

(* this does not work *)
let f (type a) : a th -> a = function
  | Thunk (type b c) (x, f : b * (b -> c)) -> f x
(*                           ^^^^^^^^^^^^
This pattern matches values of type b * (b -> c)
but a pattern was expected which matches values of type b * (b -> a)
Type c is not compatible with type a
*)
```

I think that the second example should also work: users should be able to "name explicitly" all the variables that are local to the constructor declaration.

Things get even more confusing when the variable that we would want to name is not exactly one of the rigid type variables refined by the constructor, but only a part of it:

```ocaml
type _ tho =
  | Thunk_opt : 'b * ('b -> 'c option) -> 'c option tho

let f (type a) : a tho -> a = function
  | Thunk_opt (type b c) (x, f : b * (b -> c option)) -> f x
(*                               ^^^^^^^^^^^^^^^^^^^
This pattern matches values of type b * (b -> c option)
but a pattern was expected which matches values of type b * (b -> $0 option)
Type c is not compatible with type $0
*)
```
(The error message in this example is also very unclear. Explicit names in annotations are supposed to improve the clarify of error messages, but here they are making things worse)

Currently the "named existentials" features seems to only support "pure existentials" and nothing else. But we cannot really expect users to know the internal distinctions we make between the various kind of GADT variables, so this is probably very confusing to them. Could we support the examples above?

Note: see the [current manual on named existentials](https://v2.ocaml.org/manual/gadts-tutorial.html#s:explicit-existential-name):

https://github.com/ocaml/ocaml/blob/4d932a8974d93cbd835d0bcf9f156b804ff0fcf1/manual/src/tutorials/gadtexamples.etex#L277-L283

Even in this example in the manual, I don't think it is clear to users that "all existential variables of the constructor" mean just `['a]` and not `['a; 'b]`. 

cc @garrigue

**Comment by lupjo** [comment]
Thanks @gasche for this usability feature wish, I hope it can be done in the near future

**Comment by gasche** [comment]
Hmm, don't hold your breath. Modifying the type-checker requires very technical work, the number of people with expertise enough to do this is small and they are busy, and it's very possible that this would be a fairly complex changes for reasons that I don't foresee. It may be done quickly, but it may take years or not be done at all.

**Comment by goldfirere** [comment]
As a point of comparison, Haskell allows you to match against all type variables that are in a data constructor's type. But there is some recent worry that this is the wrong design: any explicit type patern matched against a universal has to match *before* the GADT-induced equalities are available, while existential matches can happen afterwards. Because OCaml's syntax doesn't allow non-variables to be matched against, this particular issue doesn't arise, but the choice to allow both universals and existentials to behave identically has not remained obviously correct. (To be clear, I'm not making any particular recommendation here -- just saying that this is hard.)

**Comment by gasche** [comment]
> any explicit type pattern matched against a universal has to match before the GADT-induced equalities are available, while existential matches can happen afterwards. Because OCaml's syntax doesn't allow non-variables to be matched against

@goldfirere just to clarify, I am not sure what you mean by matching against a universal/existential variable (can variables themselves be pattern?). Do you have a pointer to discussions of this question with examples / more details?

Note that in OCaml we explicitly rejected the idea of comparing the type variables in the clause to the type variables in the constructor declaration; instead we reconstruct the meaning of the variables in the clause from the mandatory type annotation on the parameters of the constructor pattern. (The fact that this annotation must be at the toplevel of the "constructor pattern" is inflexible and a bit annoying.) This may be what you meant by "not allowing non-variables to be matched against".

**Comment by goldfirere** [comment]
Yes, my comment was rather mysterious. The ticket I'm thinking of in GHC-land is https://gitlab.haskell.org/ghc/ghc/-/issues/19847, but explicating it all in a way that would be meaningful to OCamlers (or Haskellers who aren't me or Simon PJ) is likely not a good use of time. A key part is that Haskell's concrete syntax for binding existentials also allows the user to write a full type, not just a type variable (e.g. `case x of MkG @(Maybe b) -> ...`). That's not possible in OCaml (e.g. `match x with MkG (type <<this must be a variable>>) -> ...`).

I didn't realize that the type annotation was mandatory here in OCaml. Haskell uses the order of the type variables in the constructor declaration.

**Comment by gasche** [comment]
I tried to use this feature again today and hit the same limitation. The code was somewhat complex, but could be simplified into the following:

```ocaml
type _ ty =
  | Int : int ty
  | Pair : 'b ty * 'c ty -> ('b * 'c) ty

(* this works *)
let rec example : type a . a ty -> a = function
| Int -> 0
| Pair (x, y) -> (example x, example y)

 (* this does not *)
let rec example : type a . a ty -> a = function
| Int -> 0
| Pair (type b c) (x, y : b ty * c ty) -> (example x, example y)
(*
                               ^^^^^^^^^^^
Error: This pattern matches values of type b ty * c ty
       but a pattern was expected which matches values of type $0 ty * $1 ty
       Type b is not compatible with type $0 
*)
```

For me personally the restriction that only "pure existentials" can be named makes the feature unusable in practice. This is a major limitation to our support for naming GADT existentials, which is important for usability.

(cc @garrigue)

**Comment by garrigue** [comment]
OK, I'll try to look into this. But this requires a change in approach, since the current code is just taking advantage that we now that those existentials, which I would call "independent" rather than "pure", are going to be introduced, and so we have just to give them a name. In your example we have to give names to existentials that are created as a result of unification, which is more tricky.

**Comment by Octachron** [comment]
Solved by #12507 .
